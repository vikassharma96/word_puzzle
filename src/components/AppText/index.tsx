import React from 'react';
import {Text, TextProps} from 'react-native';
import {useThemeColor} from '../../hooks/useTheme';
import defaultStyles from '../../styles/defaultStyles';

type AppTextProps = {
  color?: string;
} & TextProps;

const AppText = ({color, style, ...props}: AppTextProps) => {
  const defaultColor = useThemeColor('text');
  return (
    <Text
      style={[defaultStyles.text, {color: color ?? defaultColor}, style]}
      {...props}
    />
  );
};

export default AppText;
