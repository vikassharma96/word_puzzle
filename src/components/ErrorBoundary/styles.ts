import {StyleSheet} from 'react-native';
import {fontFamily, fontSizes} from '../../styles/fonts';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textStyle: {
    marginBottom: '3%',
    fontSize: fontSizes.fontSize_18,
    fontFamily: fontFamily.semiBold,
  },
});

export default styles;
