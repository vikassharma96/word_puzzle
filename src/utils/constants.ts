// Home screen constants
export const WORDS_PUZZLE = 'Words Puzzle';
export const API_ERROR_MSG = "Couldn't retrieve the categories.";
export const START = 'Start';
export const LEADERS_BOARD = 'Leaders Board';
export const ERROR_MSG = 'Please select category first';

// Puzzle screen constants
export const WRONG_TEXT = 'Sorry! \n Bad luck';
export const CORRECT_TEXT = 'Correct! \n Congratulations';
export const EARN_TEXT = 'You earn';
export const POINTS = 'Points';
export const SCORE_SUMMARY = 'Score Summay';
export const TOTAL_SCORE = 'Your total score is';
export const SHARE = 'Share';

// common constants
export const ERROR = 'Error!';
export const RETRY = 'Retry';
export const BUTTON_TEXT_OK = 'Ok';
export const BUTTON_TEXT_CANCEL = 'Cancel';
export const SKIP = 'Skip';
export const NEXT = 'Next';
export const NO_DATA_AVAILABLE = 'No data available!';
export const WORD_PUZZLE_TEXT = 'Word Puzzle!';
export const SOMETHING_WENT_WRONG = 'Something went wrong!';
export const EXIT = 'Exit';
