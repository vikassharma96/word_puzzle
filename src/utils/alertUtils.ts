import {Alert} from 'react-native';
import {BUTTON_TEXT_CANCEL, BUTTON_TEXT_OK} from './constants';

type AlertFunction = {
  title: string;
  message: string;
  singleButton?: boolean;
  positiveText?: string;
  negativeText?: string;
  cancelable?: boolean;
  positiveAction?: () => void;
  negativeAction?: () => void;
  positiveActionStyle?: any;
  negativeActionStyle?: any;
};

export const showAlert = ({
  title,
  message,
  singleButton = true,
  positiveText = BUTTON_TEXT_OK,
  negativeText = BUTTON_TEXT_CANCEL,
  cancelable = false,
  positiveAction,
  negativeAction,
  positiveActionStyle,
  negativeActionStyle,
}: AlertFunction) => {
  const actionButtons = [
    {text: positiveText, onPress: positiveAction, style: positiveActionStyle},
  ];
  if (!singleButton) {
    actionButtons.unshift({
      text: negativeText,
      onPress: negativeAction,
      style: negativeActionStyle,
    });
  }
  Alert.alert(title, message, actionButtons, {
    cancelable: cancelable,
  });
};
