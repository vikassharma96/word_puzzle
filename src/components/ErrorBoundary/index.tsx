import React, {Component, ErrorInfo, ReactNode} from 'react';
import {View} from 'react-native';
import AppText from '../AppText';
import Button from '../Button';
import {EXIT, SOMETHING_WENT_WRONG} from '../../utils/constants';
import styles from './styles';

type Props = {
  children?: ReactNode;
};

type State = {
  hasError: boolean;
};

class ErrorBoundary extends Component<Props, State> {
  public state: State = {
    hasError: false,
  };

  public static getDerivedStateFromError(_: Error): State {
    return {hasError: true};
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    // call analytics service to capture error log
    console.error(
      'ðŸš€ ~ file: ~ ErrorBoundary ~ componentDidCatch ~ error:',
      error,
      errorInfo,
    );
  }

  exitApp = () => {
    // it will crash app in production to simulate app exit behaviour
    throw {};
  };

  public render() {
    const {hasError} = this.state;
    if (hasError) {
      return (
        <View style={styles.container}>
          <AppText style={styles.textStyle}>{SOMETHING_WENT_WRONG}</AppText>
          <Button title={EXIT} onPress={this.exitApp} />
        </View>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
