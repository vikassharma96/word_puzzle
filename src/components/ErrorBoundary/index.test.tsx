import {render} from '@testing-library/react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import ErrorBoundary from '.';
import {EXIT, SOMETHING_WENT_WRONG, START} from '../../utils/constants';
import AppText from '../AppText';

describe('@components/ErrorBoundary', () => {
  const component = (
    <ErrorBoundary>
      <AppText>{START}</AppText>
    </ErrorBoundary>
  );

  it('should match snapshots', () => {
    const componentTree = renderer.create(component).toJSON();
    expect(componentTree).toMatchSnapshot();
  });

  it('should be able to handle error', () => {
    const ChildWithError = () => {
      throw new Error('Error thrown!');
    };
    const componentWithError = (
      <ErrorBoundary>
        <ChildWithError />
      </ErrorBoundary>
    );
    const {getByText} = render(componentWithError);

    const errorText = getByText(SOMETHING_WENT_WRONG);
    expect(errorText).toBeDefined();
    const exitBtn = getByText(EXIT);
    expect(exitBtn).toBeDefined();
  });
});
