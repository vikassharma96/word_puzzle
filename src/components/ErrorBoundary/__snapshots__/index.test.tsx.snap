// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/ErrorBoundary should match snapshots 1`] = `
<Text
  style={
    [
      {
        "fontFamily": "Poppins-Medium",
        "fontSize": 14,
      },
      {
        "color": "#000",
      },
      undefined,
    ]
  }
>
  Start
</Text>
`;
